### ISSUE TITLE
Root motion works unexpectedly with some models

### ISSUE THREAD

**Issue by frnsys** [godotengine/godot#75228]
### Godot version

4.0.1rc

### System information

Ubuntu 22.04, Intel Skylake GT2 [HD Graphics 520], Vulkan

### Issue description

I'm not sure if this is an engine issue or not, but with one of my character models root motion isn't working as expected.

This is the original animation:

![](https://i.stack.imgur.com/f1he8.gif)

I set up the root motion and the character stays in place as expected, but the interpreted movement shows no forward movement:

![](https://i.stack.imgur.com/90RJ3.gif)

I also notice that if I uncheck "Zero Y" in the `RootMotionView` the floor drops away:

![](https://i.stack.imgur.com/1Y6Nu.gif)

I can't figure out what the cause might be as the original animation looks correct.

### Steps to reproduce

In the attached project the `person.tscn` demonstrates the case where root motion isn't working as expected. The other two (`mixamo.tscn` and `mixamo_rotated.tscn`) are examples where it's working fine.

### Minimal reproduction project

[root_motion_issue.zip](https://github.com/godotengine/godot/files/11042097/root_motion_issue.zip)

**Comment by TokageItLab** [comment]
The problem is that your root bone has a strange Rest axis. We need to use retargeting or something so that +Z is forward like the other gltf assets.

**Comment by QuiGonMaul** [comment]
I encountered this today when testing in Godot 4.0.2.

If you've rigged and animated character in blender, then you're working in a space with the Z axis as the up and down axis.
As most users may know, it is instead the Y axis that is the up and down axis in Godot.
This hasn't been an issue up until now, because when you export from blender you can check the option "+Y Up" which orients the character correctly when importing into Godot.

However, it appears that the 'root motion view' does not interpret the converted orientation of the imported character, which is why you get the purple root motion grid moving down the Y axis - its using the character's orientation as if it was still in blender and thinks that the Z axis is still up and down.
I hope all that made sense.

This is definitely a bug with Godot 4.0 because this works perfectly in Godot 3.5
I even tested it by exporting a character without the "+Y UP" setting checked, but that results in the imported character not being oriented correctly in Godot.

**Comment by TokageItLab** [comment]
@QuiGonMaul　See also https://github.com/godotengine/godot/issues/75843#issuecomment-1501073509.

Blender has a +Y-UP coordinate space, but for cameras and bones it works like +Z-UP, so this should not be a problem as long as the model has Root bones in Blender and bone animations are used. I think the problem is probably caused when the Root motion is created as an object animation in Blender.

One currently available fix is to apply retargeting to Root bone.

**Comment by QuiGonMaul** [comment]
As far as I understand, everything else works perfectly as it also does in Godot 3.5, except the 'root motion view' purple grid.
To clarify, I'm using a blender created animated character rig that faces forward in blender's -Y axis and then exported with the +Y Up option enabled.

Without changing any settings on this imported character from blender (including suggested re-targeting fixes etc.), I'm still able to get a properly working animated character that uses root motion driven animation when running a build.
So this confirms in my experience that its only an error with the 'root motion view' itself not previewing the purple grid motion correctly, which is somewhat of a relief because its not essential to the game build and its only there as a visualisation tool - as far as I know.

**Comment by TokageItLab** [comment]
Is the your Rest of Root Bone the initial Transform?

In Godot 3.x, the motion result gets a relative value from the Rest, whereas in Godot 4.0, it gets an absolute value from the initial Transform. Therefore, Godot 4.0 should not include rotation in the Root Bone.

**Comment by QuiGonMaul** [comment]
Yes, my root bone for the rig in blender is oriented correctly with the rest of the bones (+Z axis up, and facing forward in the -Y axis).
I have successfully managed to get the character to rotate using root motion in Godot.

I should also mention that when testing the character's root motion rotation in Godot using 'root motion view', it causes the purple grid to rotate along Godot's Z axis, giving the effect of the entire floor plane spinning on its side.
To me, this suggests that Godot recognises the change in the character's orientation in a running build (essentially swapping Z for Y axis during the export/import process), but fails to recognise this change in the 'root motion view'.

**Comment by TokageItLab** [comment]
As I said above, blender uses Y-roll for bones, so the root bone must be Y up in blender.

**Comment by elXill** [comment]
I'm having this problem as well.
First my model export rig in blender: Z is forward
![Armature](https://github.com/godotengine/godot/assets/61533529/953687f2-2e80-43b1-a23f-64873caebe57)
Second my animation root motion in godot: As you can see it animation is in Z axis also movement in 3d scene also going forward which is also correct.
![Godot Root](https://github.com/godotengine/godot/assets/61533529/ed644153-3311-410e-8bdf-2e30d85ac042)
Other than this when I use code below, it also works like expected.
char_transform.global_transform.origin = char_transform.global_transform.origin +my_anim_tree.get_root_motion_position()
But after all these,In RootMotionView node Z and Y axes are mixed up same as the OP showed.

**Comment by TokageItLab** [comment]
Blender uses +Z-up for objects, but +Y-up for cameras, lights, and bones.

The axis of the root bone is preferably "aligned with the world axis of the final environment to be used". Bone axes are not converted by glTF-IO, so if they are aligned to blender's world axis, they will not be aligned to godot's world axis. So this is not a hack, but if I had to say that rather the role of the blender-glTF exporter.

**Comment by scotdoria** [comment]
I wonder if my issue has any significance with this issue. In my case my `RootMotionView` never moves even though there are root motion in my animations. I've tested in the animation preview and disabled any of the root motion track to see and the root motion animations are actually correctly imported. The weirdest part is the rotation root motion where I have an animation that should rotate by the Y-axis but instead rotates by Z-axis which makes it tumble around. It rotates properly in the animation preview but root motion interprets it differently (and my `RootMotionView` also doesn't rotate!). I could probably try and rotate the root bone to change its orientation in blender but it just doesn't feel right. My current hack is creating a new quaternion that swaps the components based on the root motion rotation.
